### ISSUE TITLE
Issue when using ORG to manage spacemacs settings

### ISSUE THREAD

**Issue by ztlevi** [syl20bnr/spacemacs#7989]
I am going to manage my spacemacs settings by org. And I load the org file in under "~/.spacemacs.d/init.el". Here is the code.

```
(package-initialize)

(setq user-emacs-directory "~/.spacemacs.d")
(require 'org-install)
(require 'ob-tangle)
(org-babel-load-file (expand-file-name "ztlevi.org" user-emacs-directory))
```

And I get the warning here.

Warning (initialization): An error occurred while loading ‘/Users/ztlevi/.emacs.d/init.el’:
Symbol's function definition is void: loop
To ensure normal operation, you should investigate and remove the
cause of the error in your initialization file.  Start Emacs with
the ‘--debug-init’ option to view a complete error backtrace.

I don't want to put my org file under ~/.emacs.d/. Is there a way to fix this?

Thanks.

**Comment by d12frosted** [comment]
You should not change `user-emacs-directory` as it breaks Spacemacs loading. Why do you want to change that? You could try following snipped to load `ztlevi.org`, but result totally depends on content of that file. And also it will end up in default version of `org-mode` to be loaded instead of the latest one.

```
(package-initialize)
(require 'org-install)
(require 'ob-tangle)
(org-babel-load-file (expand-file-name "ztlevi.org" "~/.spacemacs.d"))
```

If you wish to just combine your own configurations with Spacemacs - you have to adapt a bit. 

*Side note.* Since using literate configuration file is a popular way to manage configurations in emacs 

---

If you wish to gain more control over how and when Spacemacs is loaded you could try loading it manually. For example, I clone Spacemacs to `$HOME/.spacemacs` and in `$HOME/.emacs.d/init.el` I have following snippet to load Spacemacs.

```elisp
(defconst d12-path/xdg-config (concat (getenv "XDG_CONFIG_HOME") "/"))
(defconst d12-path/emacs-home (concat d12-path/xdg-config "emacs/"))
(defconst d12-path/spacemacs-user-config
  (concat d12-path/emacs-home "spacemacs.el"))
(defconst d12-path/spacemacs-home (concat user-home-directory ".spacemacs/"))
(setq-default spacemacs-start-directory d12-path/spacemacs-home
              dotspacemacs-filepath d12-path/spacemacs-user-config)
(load-file d12-path/spacemacs-init)
```

In short, it sets `spacemacs-start-directory` to `$HOME/.spacemacs/` and `dotspacemacs-filepath` to path of `.spacemacs` file (for me it's `$XDG_CONFIG_HOME/emacs/spacemacs.el`).

**Comment by d12frosted** [comment]
Generally I think it's a wontfix at this point. 

Anyways, closing this issue for inactivity. If anyone wishes for it to be open, just let me know.

**Comment by dudelson** [comment]
@d12frosted I'm experimenting with porting my `init.el` to a literal style using org-babel and I'm a little confused about both the problem and your proposed solution. Forgive me for not being well-versed in spacemacs internals. I tried moving a small section of `dotspacemacs/user-config` to a source code block in an org file (`init.org`) and at the end of `dotspacemacs/user-config` I placed a call to `(org-babel-load-file "./init.org")`. When I reloaded my config with `SPC f e R` I got the error

    load-file: Recursive load: "/home/david/s/dot/spacemacs/init.el", "/home/david/s/dot/spacemacs/init.el", "/home/david/s/dot/spacemacs/init.el", "/home/david/s/dot/spacemacs/init.el", "/home/david/s/dot/spacemacs/init.el", "/home/david/.spacemacs.d/init.el"

I have no idea what to make of this, and I am not certain whether it is related to the error that @ztlevi got. If it helps, my `init.el` is a symlink to the file with the path given in the error.

I remember reading somewhere that it is not possible to have a literate config with org-babel if you're using spacemacs because of the layer system and the fact that spacemacs loads a different version of org-mode than the emacs default. Is this accurate? If so, is the "side note" you gave above intended to work around this?

BTW no need to re-open the issue, just looking for some clarification. Thanks!

**Comment by d12frosted** [comment]
@dudelson 

My proposed solution was for specific case @ztlevi described. And note that I haven't tested it by myself, I just shared my thoughts on that topic 😜 I load Spacemacs manually from `init.el`, so I've also shared some parts of my old setup. Right now I have [slightly different approach](https://github.com/d12frosted/environment/blob/master/emacs/init.el), but that doesn't really matter.

Your case is different because you are not trying to load Spacemacs manually from `init.org` (though error might indicate some attempts 😸 ).

Everything you said makes sense. Move out some elisp code from `dotspacemacs/user-config` to `init.org` and load it using `(org-babel-load-file "./init.org")` in the end of the mentioned function. Except for you might want to fix the path to `init.org`, but in any case I'd rather use absolute path (with the help of `user-emacs-directory` variable). 

Are you sure that `init.org` is being loaded? I've put simple message inside of `init.org` in order to see that it is properly evaluated.

In order to help you with recursive load error, I must see what you are doing inside of `init.org` file. 

> I remember reading somewhere that it is not possible to have a literate config with org-babel if you're using spacemacs because of the layer system and the fact that spacemacs loads a different version of org-mode than the emacs default. Is this accurate? If so, is the "side note" you gave above intended to work around this?

Well. There is no general answer. Everything depends on what you are trying to do. 

- AFAIK, implementing layers in org files is not supported out of box right now. 
- Loading Spacemacs from `init.org` that is loaded from `init.el` is possible, though you must know how to do it. Depending on what you do in `init.el` you might get old version of org-mode. 
- Using `.spacemacs.org` instead of `.spacemacs` is not supported out of box, but I imagine that it's not that hard to implement. Probably. 
- Storing `user-init` and `user-config` inside of org files is supported and you described how to do it. And most importantly - it doesn't lead to org-version problems. At least should not as you call `org-mode` functions after most important stuff is initialised in Spacemacs.

---

I hope it doesn't confuse you further 😸 Let me know if you have any questions 😸

**Comment by ztlevi** [comment]
The purpose doing init.el with org for me is to nicely view the layers, while you could view it with imenu(SPC j i). Another way is to define your own layers like:
private-layer:
|-ui-layer
|-programming-layer
|-misc-layer
...
You could figure out this from my settings https://github.com/ztlevi/spacemacs-config/tree/master/layers.

BTW: I dropped the org init.el configuration because the loading time is so long!!! :(

**Comment by dudelson** [comment]
thanks for the replies. I've settled on a solution that (mostly) works for me right now, which is to load two files from `init.el`: `config.el` and `org-config.el`. These files are tangled  from `config.org` and `org-config.org` respectively. So the only difference is that I have to tangle the files manually before I reload my configuration. While this isn't optimal (there will undoubtedly be times when I go to add something to my config and forget to tangle the files and am confused when it doesn't work), it's much better than what I had before and good enough for now. I plan to revisit this in the future.

**Comment by d12frosted** [comment]
> The purpose doing init.el with org for me is to nicely view the layers, while you could view it with imenu(SPC j i).

`SPC j i` works not only in `org-mode`, but in `emacs-lisp` and other modes as well 😸

> I dropped the org init.el configuration because the loading time is so long!!! :(

Indeed, using `org-mode` init files increases loading time 😸

**Comment by dudelson** [comment]
> I dropped the org init.el configuration because the loading time is so long!!! :(
I didn't think of this before, but now it sounds like maybe I have the best of both worlds: I can view and edit my configuration in an org file, but the loading time is pretty much the same because I tangle the files manually after I edit them, and load the generated `.el` files at startup. Perhaps I will not revisit this after all.

**Comment by planigan** [comment]
I'm considering this at some point, but I hadn't considered the impact on startup time. I like @dudelson approach, but was thinking, why not find a way to auto-tangle the config files when saved. I haven't tested this, but I found this while searching for a solution to that problem: https://emacs.stackexchange.com/a/20733

**Comment by dudelson** [comment]
What a great idea! I'm adding this to my config. Thanks for the link 😸
